web-runtime/src/engine.rs

```diff
@@ -0,0 +1,143 @@
+/*
+ * Copyright 2022 Fluence Labs Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+use crate::module::MRecordTypes;
+use crate::module::MModule;
+use crate::module::MFunctionSignature;
+use crate::MResult;
+use crate::MError;
+use crate::IValue;
+use crate::IRecordType;
+
+use serde::Serialize;
+use std::collections::HashMap;
+use std::collections::hash_map::Entry;
+use std::rc::Rc;
+
+/// Represent Marine module interface.
+#[derive(PartialEq, Eq, Debug, Clone, Serialize)]
+pub struct MModuleInterface<'a> {
+    pub record_types: &'a MRecordTypes,
+    pub function_signatures: Vec<MFunctionSignature>,
+}
+
+/// The base struct of Marine, the Fluence compute runtime.
+pub struct Marine {
+    // set of modules registered inside Marine
+    modules: HashMap<String, MModule>,
+}
+
+// these methods will be used when decoupling common code from marine-runtime end web-marine-runtime
+#[allow(unused)]
+impl Marine {
+    pub fn new() -> Self {
+        Self {
+            modules: HashMap::new(),
+        }
+    }
+
+    /// Invoke a function of a module inside Marine by given function name with given arguments.
+    pub fn call<MN: AsRef<str>, FN: AsRef<str>>(
+        &mut self,
+        module_name: MN,
+        func_name: FN,
+        arguments: &[IValue],
+    ) -> MResult<Vec<IValue>> {
+        self.modules.get_mut(module_name.as_ref()).map_or_else(
```

Comment:
if you add `let module_name = module_name.as_ref()` at the beggining of this function, it won't be needed to do that every time then. And please add this to the original code.